common/pkg/capi/clustertopology/handlers/mutation/meta.go

```diff
@@ -59,6 +68,18 @@ func (mgp metaGeneratePatches) GeneratePatches(
 	resp *runtimehooksv1.GeneratePatchesResponse,
 ) {
 	clusterKey := handlers.ClusterKeyFromReq(req)
+	getCluster := func(ctx context.Context) (*clusterv1.Cluster, error) {
+		cluster := &clusterv1.Cluster{}
+		var err error
+		fn := sync.OnceFunc(func() {
+			err = mgp.cl.Get(ctx, clusterKey, cluster)
+		})
+		fn()
```

Comment:
Oh I see what you're doing now! Howeever it's still not quite right, because every call to `clusterGetter` is going to call `sync.OnceFunc`, which won't have the laziness you're after: it will always be run...

Reply:
It's all about the scoping of the `sync.Once` call and the returned variable. Here's a repro of this problem - https://go.dev/play/p/x0pnX273hhd

Reply:
I think what you want to do is like I shared in Slack, here it is again (slightly adapted), remembering we need it this way to pass the context properly for timeouts, cancellations, etc:

```
type ClusterGetterFn func(context.Context) (*capiv1.Cluster, error)

func ClusterGetterForObjectKey(cl client.Client, clusterKey client.ObjectKey) ClusterGetterFn {
    var (
        getOnce sync.Once
        cluster *capiv1.Cluster
        err error
    )
    return func(ctx context.Context) (*capiv1.Cluster, error) {
        getOnce.Do(func() {
          innerCluster := &clusterv1.Cluster{}
          innerErr := client.Get(ctx, innerCluster, clusterKey)
          if innerErr != nil {
            err = fmt.Errorf("failed to retrieve cluster: %w" innerErr)
          } else {
            cluster = innerCluster
          }
        }
        return cluster, err
    }
}
```

Reply:
Gotcha! makes sense. Thank you :bow:

Reply:
done in the latest commit